File: sliding_window/sliding_window.cpp
#include <vector>
#include <iostream>
#include <algorithm>
#include <numeric>
#include <limits>
#include <cmath>
#include "../utility/measure.hpp"

static int64_t find_max_sum_of_K_brute_force(std::vector<int>& arr, const size_t window_size)
{
    int64_t max_sum = std::numeric_limits<int64_t>::infinity();
    for(size_t i=0; i<arr.size()-window_size; i++)
    {
        int64_t sum = std::accumulate(begin(arr)+i, begin(arr)+i+window_size, 0);
        max_sum = std::max(sum, max_sum);            
    }

    return max_sum;
}

static uint64_t find_max_sum_of_K_sliding_window(std::vector<int>& arr, const size_t window_size)
{
    int64_t sum = std::accumulate(begin(arr), begin(arr)+window_size, 0);
    auto max_sum = sum;
    for(int i=0; i<arr.size()-window_size; i++)
    {
        sum -= arr[i];
        sum += arr[i+window_size];
        max_sum = std::max(sum, max_sum);
    }

    return max_sum;
}

int main(void)
{
    std::vector<int> v(99999);
    constexpr size_t WINDOW_SIZE = 999;
    srand(12314);
    for(auto& e : v)
    {
        e = random() % 10000;        
    }

    perform_print_duration("max_br: ", find_max_sum_of_K_brute_force, v, WINDOW_SIZE);
    perform_print_duration("max_sw: ", find_max_sum_of_K_sliding_window, v, WINDOW_SIZE);

    return 0;
}

File: sliding_window/sliding_window.py
import random
from datetime import datetime

def find_max_sum_of_K_brute_force(arr, window_size):
    max_sum = float('-inf')
    for i in range(len(arr)-window_size+1):
        current_sum = sum(arr[i:i+window_size])            
        max_sum = max(current_sum, max_sum)            
    return max_sum

def find_max_sum_of_K_sliding_window(arr, K):
    current_sum = sum(arr[0:K])
    
    max_sum = current_sum
    for i in range(len(arr)-K):
        current_sum -= arr[i]
        current_sum += arr[i+ K]
        max_sum = max(current_sum, max_sum)            
    return max_sum
        

random.seed(1234)
arr = random.sample(range(0, 100000), 99999)
K = 999
start = datetime.now()
max_br = find_max_sum_of_K_brute_force(arr, K)
print("max_br: {} ({})".format(max_br, (datetime.now()-start).total_seconds()))

start = datetime.now()
max_sw = find_max_sum_of_K_sliding_window(arr, K)
print("max_sw: {} ({})".format(max_sw, (datetime.now()-start).total_seconds()))